custom_components/waste_collection_schedule/waste_collection_schedule/source/avfallsapp_se.py

```diff
@@ -0,0 +1,188 @@
+from waste_collection_schedule import Collection
+from waste_collection_schedule.exceptions import (
+    SourceArgumentException,
+    SourceArgumentExceptionMultiple,
+    SourceArgumentNotFoundWithSuggestions,
+)
+
+TITLE = "Avfallsapp.se - Multi Source"
+DESCRIPTION = "Source for all Avfallsapp waste collection sources. This included multiple municipalities in Sweden."
+URL = "https://www.avfallsapp.se"
+TEST_CASES = {
+    "Söderköping - Söderköping Kommun": {
+        "api_key": "12345678",
+        "service_provider": "soggderkoping",
+    },
+}
+
+COUNTRY = "se"
+_LOGGER = logging.getLogger(__name__)
+
+ICON_RECYCLE = "mdi:recycle"
+
+# See documentatation about possible other cities and service providers
+# that look as using the same portal (sub-domain lookup).
+SERVICE_PROVIDERS = {
+    "soderkoping": {
+        "title": "Söderköping",
+        "url": "https://soderkoping.se",
+        "api_url": "https://soderkoping.avfallsapp.se/wp-json/nova/v1/",
+    },
+}
+
+EXTRA_INFO = [
+    {
+        "title": data["title"],
+        "url": data["url"],
+        "default_params": {"service_provider": provider},
+    }
+    for provider, data in SERVICE_PROVIDERS.items()
+]
+
+
+class Source:
+    def __init__(
+        self,
+        service_provider: str,
+        api_key: str | None = None,
+        street_address: str | None = None,
+    ):
+        # Raise an exception if the user did not provide akpi_key or street_address
+        if not api_key and not street_address:
+            raise SourceArgumentExceptionMultiple(
+                ["street_address", "api_key"],
+                "You must provide either a street_address or a api_key",
+            )
+        self._api_key = api_key
+        self._street_address = street_address
+        # Get the api url using the service provider
+        self._url = SERVICE_PROVIDERS.get(service_provider.lower(), {}).get("api_url")
+        if self._url is None:
+            raise SourceArgumentNotFoundWithSuggestions(
+                "service_provider",
+                service_provider,
+                SERVICE_PROVIDERS.keys(),
+            )
+        # Remove trailing slash from the url if present
+        if self._url.endswith("/"):
+            self._url = self._url[:-1]
+
+    def _register_device(self):
+        if not self._street_address:  # should not happen
+            raise Exception("street_address required to register device")
+        uuid = uuid4().hex[:16]
+
+        # if not self._api_key:
+        registerUrl = self._url + "/register"
+        params = {
+            "identifier": uuid,
+            "uuid": uuid,
+            "platform": "android",
+            "version": "2.2.0.0",
+            "os_version": "14",
+            "model": "sdk_gphone64_x86_64",
+            "test": False,
+        }
+        response = requests.post(
+            registerUrl, json=params, timeout=30, headers={"X-App-Identifier": uuid}
+        )
+        # if not response.text == "1":
```

Comment:
Not ready: Wanted to add a check that API_KEY registration worked as expected, this I can likely fix fir some more testing and to comparing to negative response.